src/hotspot/share/utilities/moveBits.hpp

```diff
@@ -26,74 +27,156 @@
 #define SHARE_UTILITIES_MOVEBITS_HPP
 
 #include "metaprogramming/enableIf.hpp"
+#include "utilities/byteswap.hpp"
+#include "utilities/debug.hpp"
 #include "utilities/globalDefinitions.hpp"
 
+#include <cstdint>
 #include <type_traits>
 
+// T reverse_bits<T>(T)
+//
+// Reverses the bits in the integral value of type T.
+
 template <typename T>
-class ReverseBitsImpl {
-  static const size_t NB = sizeof(T) * BitsPerByte;
+using CanReverseBitsImpl = std::integral_constant<bool, (std::is_integral<T>::value &&
+                                                         (sizeof(T) == 1 ||
+                                                          sizeof(T) == 2 ||
+                                                          sizeof(T) == 4 ||
+                                                          sizeof(T) == 8))>;
+
+template <typename T, size_t N = sizeof(T)>
+struct ReverseBitsImpl;
+
+template <typename T, ENABLE_IF(CanReverseBitsImpl<T>::value)>
+ALWAYSINLINE T reverse_bits(T x) {
+  return ReverseBitsImpl<T>{}(x);
```

Comment:
I think I would have preferred to have the change to bit reversal by separate from byteswap changes.
With these changes, I also think this file should be renamed to reverse_bits.hpp or something like that.
It was called moveBits because it contained both bit and byte swapping.

Reply:
Out of curiosity, why `reverse_bits.hpp` instead of `reverseBits.hpp`? I have seen a few files in `utilities/` use the former, but the rest of Hotspot seems to always use the later? Is it because they contain a single function?

I went with `reverse_bits.hpp` for now as suggested for consistency with the other bit fiddling headers in `utilities/`.

Reply:
The usual camel-case with lowercase start convention typically refers to a class name in the file.
I'm not sure snake-case for files with a single function was a good idea, even though it might have
been my fault.  But it's what we're doing (at least for now).

Reply:
Works for me, just want to make sure I am consistent with the rest.